storm-core/src/jvm/org/apache/storm/daemon/DrpcServer.java

```diff
@@ -0,0 +1,356 @@
+            throw (DRPCExecutionException) result;
+        }
+        if (result == null) {
+            throw new DRPCExecutionException("Request timed out");
+        }
+        return (String) result;
+    }
+
+    @Override
+    public void result(String id, String result) throws AuthorizationException, TException {
+        meterResultCalls.mark();
+        InternalRequest internalRequest = this.outstandingRequests.get(id);
+        if (internalRequest != null) {
+            Map<String, String> map = ImmutableMap.of(DRPCAuthorizerBase.FUNCTION_NAME, internalRequest.function);
+            checkAuthorization(authorizer, map, "result");
+            Semaphore sem = internalRequest.sem;
+            LOG.debug("Received result {} for {} at {}", result, id, System.currentTimeMillis());
+            if (sem != null) {
+                internalRequest.result = result;
+                sem.release();
+            }
+        }
+    }
+
+    @Override
+    public DRPCRequest fetchRequest(String functionName) throws AuthorizationException, TException {
+        meterFetchRequestCalls.mark();
+        Map<String, String> map = new HashMap<>();
+        map.put(DRPCAuthorizerBase.FUNCTION_NAME, functionName);
+        checkAuthorization(authorizer, map, "fetchRequest");
+        ConcurrentLinkedQueue<DRPCRequest> queue = acquireQueue(functionName);
+        DRPCRequest req = queue.poll();
+        if (req != null) {
+            LOG.debug("Fetched request for {} at {}", functionName, System.currentTimeMillis());
+            return req;
+        } else {
+            return new DRPCRequest("", "");
+        }
+    }
+
+    @Override
+    public void failRequest(String id) throws AuthorizationException, TException {
+        meterFailRequestCalls.mark();
+        InternalRequest internalRequest = this.outstandingRequests.get(id);
+        if (internalRequest != null) {
+            Map<String, String> map = new HashMap<>();
+            map.put(DRPCAuthorizerBase.FUNCTION_NAME, internalRequest.function);
+            checkAuthorization(authorizer, map, "failRequest");
+            Semaphore sem = internalRequest.sem;
+            if (sem != null) {
+                internalRequest.result = new DRPCExecutionException("Request failed");
+                sem.release();
+            }
+        }
+    }
+
+    protected ConcurrentLinkedQueue<DRPCRequest> acquireQueue(String function) {
+        ConcurrentLinkedQueue<DRPCRequest> reqQueue = requestQueues.get(function);
+        if (reqQueue == null) {
+            reqQueue = new ConcurrentLinkedQueue<>();
+            ConcurrentLinkedQueue<DRPCRequest> old = requestQueues.putIfAbsent(function, reqQueue);
+            if (old != null) {
+                reqQueue = old;
+            }
+        }
+        return reqQueue;
+    }
+
+    private void checkAuthorization(IAuthorizer aclHandler, Map mapping, String operation, ReqContext reqContext) throws AuthorizationException {
```

Comment:
This is not called in this class and has been marked private. So it can be removed I believe.